File: backend/services/mediaService.mjs
import fs from 'node:fs';
import path from 'node:path';
import multer from 'multer';

import { config } from '../config/env.mjs';
import { AppError } from '../middleware/errorHandler.mjs';
import { cleanupDeletionArtifacts } from './cleanupService.mjs';

const uploadRoot = config.uploadRoot;

const ensureDirectory = (dirPath) => {
  fs.mkdirSync(dirPath, { recursive: true });
  return dirPath;
};

export const resolveUploadKind = (value, mimeType = '') => {
  const mime = String(mimeType ?? '').trim().toLowerCase();
  const requestedKind = String(value ?? '').trim().toLowerCase();

  if (mime.startsWith('video/') || requestedKind === 'video' || requestedKind === 'videos') {
    return 'video';
  }

  return 'attachment';
};

export const getUploadFolder = (kind) => (kind === 'video' ? 'videos' : 'attachments');
export const getKindLimit = (kind) => (kind === 'video' ? 500 * 1024 * 1024 : 20 * 1024 * 1024);

export const allowedMimeTypes = {
  video: new Set(['video/mp4', 'video/webm', 'video/ogg', 'video/quicktime', 'video/x-matroska']),
  attachment: new Set([
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-powerpoint',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    'text/plain',
    'application/rtf',
    'image/png',
    'image/jpeg',
    'image/gif',
    'image/webp',
    'image/svg+xml',
  ]),
};

const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const kind = resolveUploadKind(req.body?.type, file.mimetype);
    const targetDir = ensureDirectory(path.join(uploadRoot, getUploadFolder(kind)));
    cb(null, targetDir);
  },
  filename: (req, file, cb) => {
    const extension = path.extname(file.originalname || 'upload.bin');
    const baseName = (file.originalname || 'upload')
      .replace(new RegExp(`${extension.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}$`), '')
      .replace(/[^a-zA-Z0-9._-]/g, '_')
      .slice(0, 80) || 'upload';
    const uniqueSuffix = `${Date.now()}-${Math.round(Math.random() * 1e9)}`;
    cb(null, `${baseName}-${uniqueSuffix}${extension}`);
  },
});

export const uploadMedia = multer({
  storage,
  limits: { fileSize: 500 * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    const kind = resolveUploadKind(req.body?.type, file.mimetype);
    const allowed = allowedMimeTypes[kind] ?? allowedMimeTypes.attachment;

    if (allowed.has(file.mimetype)) {
      cb(null, true);
      return;
    }

    cb(new AppError(`Invalid file type for ${kind}.`, 415, { receivedType: file.mimetype }));
  },
});

ensureDirectory(path.join(uploadRoot, 'videos'));
ensureDirectory(path.join(uploadRoot, 'attachments'));

export async function processMediaUpload(req) {
  if (!req.file) {
    throw new AppError('No file uploaded.', 400, { field: 'file' });
  }

  const detectedKind = resolveUploadKind(req.body?.type, req.file.mimetype);
  const maxFileSize = getKindLimit(detectedKind);

  if (req.file.size > maxFileSize) {
    fs.unlinkSync(req.file.path);
    throw new AppError('File exceeds the allowed size limit.', 413, {
      maxFileSize,
      receivedFileSize: req.file.size,
    });
  }

  const folder = getUploadFolder(detectedKind);
  const url = `/uploads/${folder}/${req.file.filename}`;

  return {
    url,
    folder,
    type: detectedKind,
    fileName: req.file.filename,
    originalName: req.file.originalname,
  };
}

export async function processDeletion(req) {
  const entity = req.body?.entity ?? req.body?.course ?? req.body?.lesson ?? req.body ?? null;

  if (!entity || typeof entity !== 'object') {
    throw new AppError('Missing entity payload for deletion cleanup.', 400, { body: req.body });
  }

  const cleanupResult = cleanupDeletionArtifacts(entity, {
    projectRoot: config.projectRoot,
    databaseStore: req.app.locals.databaseStore ?? {},
  });

  return cleanupResult;
}

File: backend_fastapi/app/services/media_service.py
from __future__ import annotations

import hashlib
import mimetypes
import os
from pathlib import Path
from typing import Any

from fastapi import HTTPException, UploadFile, status

from app.services.cleanup_service import cleanup_deletion_artifacts

PROJECT_ROOT = Path(__file__).resolve().parents[2]
UPLOAD_ROOT = PROJECT_ROOT / "uploads"
UPLOAD_ROOT.mkdir(parents=True, exist_ok=True)

ALLOWED_MIME_TYPES = {
    "video": {"video/mp4", "video/webm", "video/ogg", "video/quicktime", "video/x-matroska"},
    "attachment": {
        "application/pdf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/vnd.ms-powerpoint",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "text/plain",
        "application/rtf",
        "image/png",
        "image/jpeg",
        "image/gif",
        "image/webp",
        "image/svg+xml",
    },
}


def resolve_upload_kind(value: str | None, mime_type: str = "") -> str:
    requested_kind = str(value or "").strip().lower()
    mime = (mime_type or "").strip().lower()
    if mime.startswith("video/") or requested_kind in {"video", "videos"}:
        return "video"
    return "attachment"


def get_upload_folder(kind: str) -> str:
    return "videos" if kind == "video" else "attachments"


def get_kind_limit(kind: str) -> int:
    return 500 * 1024 * 1024 if kind == "video" else 20 * 1024 * 1024


async def ensure_upload_directory(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


async def process_media_upload(file: UploadFile, type_value: str | None = None, project_root: str | Path | None = None) -> dict[str, Any]:
    if file is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail={"error": "No file uploaded.", "details": {"field": "file"}})

    raw_kind = resolve_upload_kind(type_value, file.content_type or "")
    max_size = get_kind_limit(raw_kind)

    if file.size and file.size > max_size:
        raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, detail={"error": "File exceeds the allowed size limit.", "details": {"maxFileSize": max_size, "receivedFileSize": file.size}})

    allowed = ALLOWED_MIME_TYPES.get(raw_kind, ALLOWED_MIME_TYPES["attachment"])
    mime_type = file.content_type or "application/octet-stream"
    if mime_type not in allowed:
        raise HTTPException(status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE, detail={"error": f"Invalid file type for {raw_kind}.", "details": {"receivedType": mime_type}})

    workspace_root = Path(project_root) if project_root is not None else PROJECT_ROOT
    target_dir = workspace_root / "uploads" / get_upload_folder(raw_kind)
    await ensure_upload_directory(target_dir)

    original_name = file.filename or "upload.bin"
    extension = Path(original_name).suffix or mimetypes.guess_extension(mime_type) or ".bin"
    safe_name = Path(original_name).stem.replace("/", "_").replace("\\", "_")
    safe_name = "".join(ch for ch in safe_name if ch.isalnum() or ch in {"_", "-", "."}) or "upload"
    digest = hashlib.sha1(os.urandom(8)).hexdigest()[:12]
    stored_name = f"{safe_name}-{digest}{extension}"
    file_path = target_dir / stored_name

    content = await file.read()
    file_path.write_bytes(content)

    route = f"/uploads/{get_upload_folder(raw_kind)}/{stored_name}"
    return {
        "url": route,
        "folder": get_upload_folder(raw_kind),
        "type": raw_kind,
        "file_name": stored_name,
        "original_name": original_name,
    }


async def process_deletion(entity: dict[str, Any] | None, project_root: str | Path | None = None, database_store: dict[str, Any] | None = None) -> dict[str, Any]:
    if entity is None or not isinstance(entity, dict):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail={"error": "Missing entity payload for deletion cleanup.", "details": {"body": entity}})

    workspace_root = Path(project_root) if project_root is not None else PROJECT_ROOT
    return cleanup_deletion_artifacts(entity, project_root=workspace_root, database_store=database_store)
